<table style="width:100%; font-size:11pt; border-collapse:collapse">
    <tr>
        <td colspan="2"
            style="border: 1px #0098cd solid;
                   background-color:#E6F4F9;
                   color:#0098CD;
                   text-align:center;
                   font-weight:bold;
                   padding:8px;">
            Universidad de Oriente
        </td>
    </tr>
    <tr>
        <td style="border: 1px #0098cd solid;
                   background-color:#E6F4F9;
                   color:#0098CD;
                   width:50%;
                   text-align:center;
                   padding:6px;">
                            Machine Learning
        </td>
        <td style="border: 1px #0098cd solid;
                   background-color:#E6F4F9;
                   color:#0098CD;
                   width:50%;
                   text-align:center;
                   padding:6px;">
            Clase 4: Estadística robusta - Continuación
        </td>
    </tr>
    <tr>
 

</table>

In [2]:
import pandas as pd
import numpy as np
from pandas.core.common import flatten
from plotnine import *
from array import *
import scipy.stats as stats
import math
import matplotlib as mpl
import matplotlib.pyplot as plt
import statsmodels.api as sm
from sklearn import linear_model
import statsmodels.formula.api as smf
import statsmodels.stats.api as sms

## Ejemplo 1

Se tiene la siguiente muestra $X=\{4,3,11,5,4,50,6,8,7,9 \}$, en el que el valor 50 es un outlier

a) Calcular la media recortada de la muestra al 20\%, al 10\% y calcula la centrimedia

b) Calcular la media winsorizada al 10 %.

c) Calcular la trimedia

d) Compara los valores obtenidos en los apartados a), b) y c) con la media muestral.

e) Calcula el MAD y comparalo con la desviación típica

a) Para calcular las medias recortadas, se eliminan los datos una vez ordenados por cada extremo y se calcula la media.

In [2]:
#Definir la muestra
X= np.array([4, 3, 11, 5, 4, 50, 6, 8, 7, 9])


#Ordenar los valores de la muestra
X.sort()
print(X)

[ 3  4  4  5  6  7  8  9 11 50]


In [3]:
# Media recortada al 20%, eliminar 0.20(10)=2 de cada lado
X1 = X[2:8]
media_recortada_20 = np.sum(X1)/6

print("La media recortada al 20%% es %s" %media_recortada_20)

La media recortada al 20% es 6.5


In [4]:
# Media recortada al 10%, eliminar 0.10(10)=1 de cada lado
X2 = X[1:9]
media_recortada_10 = np.sum(X2)/8

print("La media recortada al 10%% es %s" %media_recortada_10)

La media recortada al 10% es 6.75


In [5]:
#Cálculo de la centrimedia, alfa=25%, 0.25(10)=2.5, eliminar 2 de cada lado
# Y ponderar el primero y último con p=1+[alfa*n]-alfa*n = 1+[0.25(10)]-0.25*10= 1+2-2.5=0.5-+278
X3 = X[2:8] #Eliminar 2 de cada lado
X3[0] = 0.5*X3[0] #Ponderar el primer valor
X3[-1]=0.5*X3[-1] #Ponderar el último valor

centrimedia = np.sum(X3)/6

print("La centrimedia es %s" %centrimedia)

La centrimedia es 5.333333333333333


b) Para obtener la media winsorizada, en lugar de eliminar las observaciones, se sustituyenpor los últimos valores. Es análogo a las medias recortadas excepto en que las
puntuaciones que se eliminaban ahora se sustituyen por los valores menor y mayor que quedan para el cómputo de la media winsorizada.

In [8]:
#Definir la muestra
X = np.array([4, 3, 11, 5, 4, 50, 6, 8, 7, 9])

#Media winsorizada al 10%

#Ordenar los valores de la muestra
X_ord = X.sort()
# print(X_ord)

# quitar valores recortados 

#Reemplazar el primer y ultimo valor por el más cercano
X[0] = X[1]
X[-1] = X[-2]

# print(W1)
print("La media winsorizada es", sum(X)/10)

La media winsorizada es 6.9


c) La trimedia se obtiene ponderando la mediana y el primer y tercer cuartil según
la fórmula:

$$\text{trimedia} = \dfrac{\text{cuartil 1}+2 \text{cuartil 2} +\text{cuartil 3}}{4}$$

In [13]:
#Definir la muestra
X = np.array([4, 3, 11, 5, 4, 50, 6, 8, 7, 9])

#Ordenar los valores de la muestra
X.sort()
q1_x = np.quantile(X, 0.25) #Cuartil 1
#Cuartil 2
me_x = np.quantile(X, 0.50) #Cuartil 2
#Cuartil 3
q3_x = np.quantile(X, 0.75) #Cuartil 3

trimedia = (q1_x+2*me_x+q3_x)/4
print("La trimedia es %s" %trimedia)

La trimedia es 6.5


d) Se obtiene la media muestral

In [14]:
#Definir la muestra
X = np.array([4, 3, 11, 5, 4, 50, 6, 8, 7, 9])

#Ordenar los valores de la muestra
X.sort()
print("La media muestral es", X.mean())

La media muestral es 10.7


La media muestral es 10.7, que es mucho más sensible al outlier de 50, que los métodos
robustos obtenidos en los apartados a, b y c.

e) Para calcular el MAD es necesario calcular la mediana de la diferencia absoluta de
cada valor con la mediana de la muestra.

In [15]:
print("La mediana de la muestra vale me_x = ", me_x)

La mediana de la muestra vale me_x =  6.5


In [17]:
#Construir vector diferencias absolutas
AD = abs(X-me_x) #Vector de diferencias de la muestra y mediana
MAD = np.quantile(AD, 0.5) #interpolation = "midpoint"
print("El MAD vale", MAD)

El MAD vale 2.5


In [18]:
print("La desviación típica es", round(X.std(), 1))

La desviación típica es 13.3
